core/blockchain.go

```diff
@@ -1407,6 +1407,9 @@ var lastWrite uint64
 // but does not write any state. This is used to construct competing side forks
 // up to the point where they exceed the canonical total difficulty.
 func (bc *BlockChain) writeBlockWithoutState(block *types.Block, td *big.Int) (err error) {
+	if bc.insertStopped() {
+		return errInsertionInterrupted
+	}
```

Comment:
We should put the check after the `wg`. Otherwise the blockchain can be stopped after the check and before the wg operations. So we will still operate the closed database.

Reply:
Myeah, the thing is, that `bc.wg.Add` is a bit YOLO. We're not doing it by the book, and I preferred to 

```
// Note that calls with a positive delta that occur when the counter is zero
// must happen before a Wait. Calls with a negative delta, or calls with a
// positive delta that start when the counter is greater than zero, may happen
// at any time.
// Typically this means the calls to Add should execute before the statement
// creating the goroutine or other event to be waited for.
// If a WaitGroup is reused to wait for several independent sets of events,
// new Add calls must happen after all previous Wait calls have returned.
```